### PR TITLE
Support for ServoSession

### PR DESCRIPTION
**Pull request by paulrouget**
This is an early PR to support Servo. It toggles from Gecko to Servo when the homebutton is pressed (just as a temporary trigger). It overrides GeckoSession and a few other classes.

I'd like to get an early feedback to know if this approach is sensible.

@MortimerGoro Can I ask you to take a look?

### CODE REVIEW

**Comment by paulrouget** [comment]
Note: I can't extend GeckoDisplay and PanZoomController. See https://bugzilla.mozilla.org/show_bug.cgi?id=1489933

**Comment by paulrouget** [comment]
Can we make Servo support non optional? The AAR would be pulled from download.servo.org. Or at least make sure the CI builds and test Servo support?

If we don't build Servo automatically, I'm afraid Servo support will break every other day.

**Comment by paulrouget** [comment]
> is Maven deploy possible?

Yes. I'll see if we can use is maven.m.o or if we should setup something under servo.org.

**Comment by paulrouget** [comment]
Depends on https://bugzilla.mozilla.org/show_bug.cgi?id=1489933 and https://bugzilla.mozilla.org/show_bug.cgi?id=1494913

**Comment by paulrouget** [comment]
@MortimerGoro this is ready for review but not 100% ready to land. Waiting on the 2 bugs mentioned above. But I would appreciate an early review.

File: app/src/common/shared/org/mozilla/servo/ServoDisplay.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    @Override
+    public void surfaceChanged(final Surface surface, final int width, final int height) {
+        mServoSession.onSurfaceReady(surface, width, height);
+    }
+
+    @Override
+    public void surfaceDestroyed() {
+        mServoSession = null;
```
Comment (MortimerGoro) [inline app/src/common/shared/org/mozilla/servo/ServoDisplay.java]: on GeckoView `surfaceDestroyed` pauses the compositor internally. We should so something similar in Servo. It's used to pause the compositor while in WebVR, or also when the activity goes to background.

File: app/src/main/res/layout/navigation_bar.xml [hunk 2]
```diff
@@ -40,6 +40,13 @@
                 android:layout_marginStart="10dp"
                 android:src="@drawable/ic_icon_home" />
 
+            <org.mozilla.vrbrowser.ui.UIButton
```
Comment (MortimerGoro) [inline app/src/main/res/layout/navigation_bar.xml]: this is for testing right?

**Comment by MortimerGoro** [comment]
There are still some TODOs (e.g. handle all the kind of motion events correctly, pausing the compositor, ...) but code looks good on a high level

File: app/src/main/res/layout/navigation_bar.xml (hunk 2 above)
Comment (paulrouget) [reply app/src/main/res/layout/navigation_bar.xml]: Well, we want this to land. It only shows when the pref is on.
We will think of a better UI later I guess.

**Comment by paulrouget** [comment]
For the TODO in the panzoom controller, I was thinking about doing that in a follow-up PR.
For the pausing the compositor, or anything that has to do with stopping Servo, I was also plannign to do that later (in https://github.com/servo/servo/issues/21834)

**Comment by paulrouget** [comment]
@MortimerGoro I filed follow-up issues for all the FIXME. Once this lands, I will also file a follow-up for the mouse events.

There are only 2 "FIXME BEFORE LANDING" left that I will fix once the maven repo is published.

Can I get a formal review?

**Comment by paulrouget** [comment]
Before landing, we need to:
- update to GV to 64.0.20181003
- publish the maven repo and update this PR

**Comment by jdm** [comment]
http://servo-builds.s3.amazonaws.com/nightly/maven/org/mozilla/servoview/servoview-armv7/maven-metadata.xml now exists and will be updated as part of the nightly build.

**Comment by paulrouget** [comment]
Maven repo is available and GV has been updated.

This is ready to land.

**Comment by paulrouget** [comment]
@MortimerGoro we are going to target the Oculus store for now. We will implement that expansion system later.

> Tested some pages such as youtube, sketchfab and they didn't work. After loading sketchfab I couldn't navigate to more websites. Are DOM images broken? I can't see them in the home URL or in a google search.

We are aware that many websites are broken. We will follow up on these issues after this lands. Our plan is to enable the Servo button only for websites that we know work correctly.

**Comment by paulrouget** [comment]
@MortimerGoro I have addressed your comments. Can you take a look and confirm that this is what you wanted.

It's not ready for review because I started getting an "incompatible app" error. I have no idea why. This is unrelated to code change afaict. Still investigating.

**Comment by paulrouget** [comment]
I finally worked around the "incompatible app" error.

This is ready for review.

I don't have an Oculus Go device, so please someone give it a try (we might run into an issue where the oculus androidmanifest is not included in the apk).

File: servo/build.gradle [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+          implementation files(aar)
+        }
+    } else {
+        throw new GradleException('Failed to find ServoView AAR at: ' + gradle.servoViewLocal)
+    }
+} else {
+    dependencies {
+        // To see what the latest servoview version is go here:
+        // https://download.servo.org/nightly/maven/org/mozilla/servoview/servoview-armv7/maven-metadata.xml
+        implementation 'org.mozilla.servoview:servoview-armv7:0.0.1.20181005.caa4d190af'
```
Comment (jdm) [inline servo/build.gradle]: Please use 0.0.1.20181015.532775 to pull in the latest fixes for oculus.

**Comment by jdm** [comment]
Looks like our generated maven manifest doesn't contain valid data, so that will have to wait.

**Comment by jdm** [comment]
0.0.1.20181015.5327758 should now work.

**Comment by MortimerGoro** [comment]
@jdm What does 0.0.1.20181015.5327758 fix?

**Comment by jdm** [comment]
@MortimerGoro Rendering of images without alpha channels (and many three.js examples), as well as reloading pages containing images.

**Comment by jdm** [comment]
I've tried debug and release servo builds, and I get this:
```
0-15 18:00:31.284   984  1243 I ActivityManager: Start proc 31704:org.mozilla.vrbrowser/u0a62 for activity org.mozilla.vrbrowser/.VRBrowserActivity
10-15 18:00:31.310   984  1288 D ActivityManager: RestoreAudioRecord called for pkg:org.mozilla.vrbrowser
10-15 18:00:31.311   984  1288 D ActivityManager: restoring op audio record to 1 for pkg:org.mozilla.vrbrowser
10-15 18:00:31.341 31278 31328 E VRB     : ======> Found PID 31704
10-15 18:00:31.397 31704 31723 E GeckoLibLoad: Load sqlite start
10-15 18:00:31.402 31704 31704 W Settings: Setting animator_duration_scale has moved from android.provider.Settings.System to android.provider.Settings.Global, returning read-only global URI.
10-15 18:00:31.409 31704 31723 W GeckoLinker: /data/app/org.mozilla.vrbrowser-1/base.apk!/assets/armeabi-v7a/libnss3.so: Relocation to NULL @0x0012ffd0 for symbol "__cxa_type_match"
10-15 18:00:31.409 31704 31723 E GeckoLibLoad: Load sqlite done
10-15 18:00:31.409 31704 31723 E GeckoLibLoad: Load nss start
10-15 18:00:31.409 31704 31723 E GeckoLibLoad: Load nss done
10-15 18:00:31.409 31704 31723 W GeckoLinker: Couldn't map /data/user/0/org.mozilla.vrbrowser/cache/libxul.so.crc to validate checksum
10-15 18:00:31.410 31704 31704 E VRB     : in onCreate
10-15 18:00:31.428 31704 31704 D AndroidRuntime: Shutting down VM
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: >>> REPORTING UNCAUGHT EXCEPTION FROM THREAD 1 ("main")
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: java.lang.IllegalArgumentException: Unable to find native library main using classloader: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/org.mozilla.vrbrowser-1/base.apk"],nativeLibraryDirectories=[/data/app/org.mozilla.vrbrowser-1/lib/arm, /data/app/org.mozilla.vrbrowser-1/base.apk!/lib/armeabi-v7a, /system/lib, /vendor/lib]]]
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: 	at android.app.NativeActivity.onCreate(NativeActivity.java:168)
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: 	at org.mozilla.vrbrowser.PlatformActivity.onCreate(PlatformActivity.java:27)
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: 	at org.mozilla.vrbrowser.VRBrowserActivity.onCreate(VRBrowserActivity.java:130)
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: 	at android.app.Activity.performCreate(Activity.java:6682)
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1119)
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2633)
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2741)
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: 	at android.app.ActivityThread.-wrap12(ActivityThread.java)
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1478)
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: 	at android.os.Handler.dispatchMessage(Handler.java:102)
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: 	at android.os.Looper.loop(Looper.java:154)
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: 	at android.app.ActivityThread.main(ActivityThread.java:6143)
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: 	at java.lang.reflect.Method.invoke(Native Method)
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:889)
10-15 18:00:31.429 31704 31704 E GeckoCrashHandler: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
```
The non-servo builds work fine.

**Comment by paulrouget** [comment]
Pretty sure it's missing manifest. It's not as easy as for googlevr. Looking at it.

**Comment by paulrouget** [comment]
I tested with a release build (googlevr) and it works.

**Comment by jdm** [comment]
The latest changes work on Oculus.

**Pull request merged**